nucypher/blockchain/eth/agents.py

```diff
@@ -322,6 +283,130 @@ def collect_refund(self, policy_id: str, author_address: str) -> str:
+        self.__read_proxy()
+        super().__init__(blockchain=self.blockchain, contract=self.principal_contract, *args, **kwargs)
+
+        self.token_agent = NucypherTokenAgent(blockchain=self.blockchain)
+
+    def __read_proxy(self):
+        self.__proxy_agent = self.UserEscrowProxyAgent(blockchain=self.blockchain)
+        contract = self.__proxy_agent._generate_beneficiary_agency(principal_address=self.principal_contract.address)
+        self.__proxy_contract = contract
+
+    def __fetch_principal_contract(self, contract_address: str = None) -> None:
+        """Fetch the UserEscrow deployment directly from the AllocationRegistry."""
+        if contract_address is not None:
+            contract_data = self.__allocation_registry.search(contract_address=contract_address)
+        else:
+            contract_data = self.__allocation_registry.search(beneficiary_address=self.beneficiary)
+        address, abi = contract_data
+        principal_contract = self.blockchain.interface.w3.eth.contract(abi=abi,
+                                                                       address=address,
+                                                                       ContractFactoryClass=Contract)
+        self.__principal_contract = principal_contract
+
+    def __set_owner(self) -> None:
+        owner = self.owner
+        self.__beneficiary = owner
+
+    def __read_principal(self, contract_address: str = None) -> None:
+        self.__fetch_principal_contract(contract_address=contract_address)
+        self.__set_owner()
+
+    @property
+    def owner(self):
+        owner = self.principal_contract.functions.owner().call()
+        return owner
+
+    @property
+    def beneficiary(self):
+        return self.__beneficiary
+
+    @property
+    def proxy_contract(self):
+        """Directly reference the beneficiary's deployed contract instead of the proxy contracts's interface"""
+        if self.__proxy_contract is NO_CONTRACT_AVAILABLE:
+            raise RuntimeError("{} not available".format(self.registry_contract_name))
+        return self.__proxy_contract
+
+    @property
+    def principal_contract(self):
+        """Directly reference the beneficiary's deployed contract instead of the proxy contracts's interface"""
+        if self.__principal_contract is NO_CONTRACT_AVAILABLE:
+            raise RuntimeError("{} not available".format(self.registry_contract_name))
+        return self.__principal_contract
+
+    @property
+    def allocation(self):
+        return self.principal_contract.functions.lockedValue().call()
+
+    @property
+    def end_timestamp(self):
+        return self.principal_contract.functions.endLockTimestamp().call()
+
+    @property
+    def locked_tokens(self) -> int:
+        """Returns the amount of tokens this miner has locked for the beneficiary."""
+        return self.principal_contract.functions.getLockedTokens().call()
+
+    def lock(self, amount: int, periods: int) -> str:
```

Comment:
it can confuse, maybe change name of the method or add comment, because `lock` is not directly related to the previous method `locked_tokens`

Reply:
:heavy_check_mark: